### PR TITLE
Transition Interpolator

### PR DESCRIPTION
**Pull request by okiwi6**
## Introduced Changes

Adds a `TransitionInterpolator` which interpolates between only two keys, replacing `LinearInterpolator`.
For the `dispatching_interpolator`, the hard coded value of `1s` is removed and replaced by a calculation of
transition time using the `maximum_joint_velocities` (which is a new parameter).
This theoretically allows for a faster transition if the robot is already close to the target position.

Fixes #

## ToDo / Known Issues
* The `maximum_joint_velocities` are chosen pretty arbitrarily

## Ideas for Next Iterations (Not This PR)

*If there are some improvements that could be done in a next iteration, describe them here.*

## How to Test

* Build and upload
* A new parameter `maximum_joint_velocities` is available
* The dispatching interpolator movements should look the same, the duration can be different however
* The arm movement while walking should be unchanged

### CODE REVIEW

File: crates/types/src/joints.rs [hunk 1]
```diff
@@ -29,6 +29,10 @@ impl HeadJoints {
             pitch: value,
         }
     }
+
+    pub fn raw(&self) -> Vec<f32> {
```
Comment (schluis) [inline crates/types/src/joints.rs:33]: Finde die Bezeichnung unklar. Vielleicht irgendwas wie as_vec oder so?
Gilt natürlich auch für alle anderen Stellen

File: crates/types/src/joints.rs [hunk 2]
```diff
@@ -51,6 +55,17 @@ impl Mul<f32> for HeadJoints {
     }
 }
 
+impl Div<HeadJoints> for HeadJoints {
+    type Output = HeadJoints;
+
+    fn div(self, right: HeadJoints) -> Self::Output {
```
Comment (schluis) [inline crates/types/src/joints.rs:61]: Wieso right? Hätte eher sowas wie other gewählt. Gibt es da eine convention?
Gilt natürlich auch für alle anderen Stellen

File: etc/configuration/default.json
Comment (schluis) [inline etc/configuration/default.json:1]: Formatierung scheint falsch zu sein

**Review by schluis — CHANGES_REQUESTED** [review]
Das sieht schon sehr nice aus!

File: crates/types/src/joints.rs (hunk 2 above)
Comment (knoellle) [reply crates/types/src/joints.rs:61]: Yes, operands of binary operators are usually called left-hand side (lhs) and right-hand side (rhs).
This makes the order more explicit compared to using "other", which is especially important for non commutative operations such as division.

Comment (h3ndrk) [reply crates/types/src/joints.rs:61]: Yes, prefer `other` or `left` or `right`

File: etc/configuration/default.json
Comment (okiwi6) [reply etc/configuration/default.json:1]: Fixed

File: crates/types/src/joints_velocity.rs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+#[derive(Clone, Copy, Debug, Default, Serialize, Deserialize, SerializeHierarchy)]
+pub struct JointsVelocity {
+    joint_velocities: Joints,
+}
+
+impl Div<JointsVelocity> for Joints {
+    type Output = JointsTime;
+
+    fn div(self, rhs: JointsVelocity) -> Self::Output {
```
[the remark this answers is not in the document]
Comment (okiwi6) [reply crates/types/src/joints_velocity.rs]: Ups, ist jetzt `right`

File: crates/control/src/transition_interpolator.rs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+use color_eyre::{eyre::Context, Result};
+
+#[derive(Clone, Default, Debug, Serialize, Deserialize)]
+pub struct TransitionInterpolator<T: Debug + Interpolate<f32>> {
+    #[serde(skip)]
+    interpolator: SplineInterpolator<T>,
+}
+
+impl TransitionInterpolator<Joints> {
+    pub fn try_new_with_maximum_velocity(
```
[the remark this answers is not in the document]
Comment (okiwi6) [reply crates/control/src/transition_interpolator.rs]: Würde ich wiedersprechen, in der std library wird auch meistens `try` verwendet, wenn es ein `Result` zurückgibt.

File: crates/control/src/motion/dispatching_interpolator.rs [hunk 5]
```diff
@@ -2,14 +2,16 @@ use std::time::Duration;
...
-    BodyJointsCommand, CycleTime, HeadJoints, Joints, JointsCommand, LinearInterpolator,
+    BodyJointsCommand, CycleTime, HeadJoints, Joints, JointsCommand, JointsVelocity,
     MotionSafeExits, MotionSelection, MotionType, SensorData,
 };
 
+use crate::transition_interpolator::TransitionInterpolator;
+
 pub struct DispatchingInterpolator {
-    interpolator: LinearInterpolator<Joints>,
+    interpolator: TransitionInterpolator<Joints>,
```
[the remark this answers is not in the document]
Comment (okiwi6) [reply crates/control/src/motion/dispatching_interpolator.rs]: Stimmt eigentlich :D, habe `TransitionInterpolator` entfernt und die Methoden zum `SplineInterpolator` geschoben

[the remark this answers is not in the document]
Comment (schluis) [reply crates/control/src/motion/dispatching_interpolator.rs]: Nice

File: crates/control/src/motion/dispatching_interpolator.rs [hunk 6]
```diff
@@ -53,11 +58,14 @@ impl DispatchingInterpolator {
...
 
-    pub fn cycle(&mut self, context: CycleContext) -> Result<MainOutputs> {
+    pub fn cycle(&mut self, mut context: CycleContext) -> Result<MainOutputs> {
         context.motion_safe_exits[MotionType::Dispatching] = false;
 
         let currently_active = context.motion_selection.current_motion == MotionType::Dispatching;
         if !currently_active {
+            context
+                .transition_time
+                .fill_if_subscribed(|| Duration::ZERO);
```
[the remark this answers is not in the document]
Comment (schluis) [reply crates/control/src/motion/dispatching_interpolator.rs]: Top

File: crates/control/src/transition_interpolator.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (schluis) [reply crates/control/src/transition_interpolator.rs]: okay, in Ordnung

File: crates/control/src/transition_interpolator.rs [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        Self::try_new_timed(
+            current_position,
+            target_position,
+            maximum_time_to_completion,
+        )
+    }
+}
+
+impl<T: Debug + Interpolate<f32>> TransitionInterpolator<T> {
+    pub fn try_new_timed(
```
[the remark this answers is not in the document]
Comment (schluis) [reply crates/control/src/transition_interpolator.rs]: Na gut

File: etc/configuration/default.json
Comment (schluis) [reply etc/configuration/default.json:1]: Das war mir gar nicht aufgefallen, das meinte ich nicht

File: crates/types/src/joints.rs (hunk 1 above)
Comment (schluis) [reply crates/types/src/joints.rs]: Top

File: crates/types/src/joints.rs [hunk 8]
```diff
@@ -500,6 +586,41 @@ where
...
+//                 head: self.head / right.head,
+//                 left_arm: self.left_arm / right.left_arm,
+//                 right_arm: self.right_arm / right.right_arm,
+//                 left_leg: self.left_leg / right.left_leg,
+//                 right_leg: self.right_leg / right.right_leg,
+//             },
+//         }
+//     }
+// }
+
```
Comment (schluis) [inline crates/types/src/joints.rs:623]: Did you forget this?

**Review by schluis — CHANGES_REQUESTED** [review]

File: crates/types/src/joints.rs (hunk 8 above)
Comment (okiwi6) [reply crates/types/src/joints.rs:623]: Oops, removed

File: crates/control/src/spline_interpolator.rs [hunk 9]
```diff
@@ -163,4 +207,12 @@ impl SplineInterpolator {
     pub fn reset(&mut self) {
         self.current_time = Duration::ZERO;
     }
+
+    pub fn duration(&self) -> Duration {
+        self.end_time
+    }
+
+    pub fn passed_duration(&self) -> Duration {
+        self.current_time
```
[the remark this answers is not in the document]
Comment (okiwi6) [reply crates/control/src/spline_interpolator.rs:216]: Done

File: etc/configuration/default.json
Comment (schluis) [reply etc/configuration/default.json:1]: Top

File: crates/control/src/spline_interpolator.rs [hunk 10]
```diff
@@ -1,15 +1,26 @@
...
     current_time: Duration,
     end_time: Duration,
 }
 
+impl<T: Debug + Interpolate<f32>> Default for SplineInterpolator<T> {
+    fn default() -> Self {
+        Self {
+            spline: Spline::from_vec(vec![]),
+            current_time: Duration::ZERO,
+            end_time: Duration::ZERO,
```
[the remark this answers is not in the document]
Comment (okiwi6) [reply crates/control/src/spline_interpolator.rs]: done

[the remark this answers is not in the document]
Comment (schluis) [reply crates/control/src/spline_interpolator.rs]: nice

**Review by schluis — DISMISSED** [review]
Nice, thank you!

**Comment by okiwi6** [comment]
I'm sorry we just noticed a bug during a testgame that may come from this PR, I have to investigate that :sweat_smile:

**Comment by okiwi6** [comment]
The bug is "fixed" now

**Review by schluis — APPROVED** [review]
Wilde Sache

**Pull request merged**